sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/interpreter/BeamSqlFnExecutor.java

```diff
@@ -503,13 +514,37 @@ private static boolean isDateNode(SqlTypeName type, Object value) {
   public void prepare() {}
 
   @Override
-  public List<Object> execute(
-      Row inputRow, BoundedWindow window, ImmutableMap<Integer, Object> correlateEnv) {
-    List<Object> results = new ArrayList<>();
-    for (BeamSqlExpression exp : exps) {
-      results.add(exp.evaluate(inputRow, window, correlateEnv).getValue());
+  public @Nullable List<Object> execute(
+      Row inputRow,
+      BoundedWindow window,
+      ImmutableMap<Integer, Object> correlateEnv,
+      ImmutableMap<Integer, Object> localRefEnv) {
+
+    int nextExprId = 0;
+    for (BeamSqlExpression expr : exprs) {
+      localRefEnv =
```

Comment:
This is a hot path and I don't think `ImmutableMap` is intelligent about updates, especially given the lack of an immutable `put`. Anyhow I think this can be an array based on how the variables are packed.

Reply:
Done